capsules/extra/src/i2c_nmea.rs

```diff
@@ -0,0 +1,169 @@
+        }
+
+        Ok(())
+    }
+}
+
+impl<'a, I: I2CDevice> I2CClient for I2cNmea<'a, I> {
+    fn command_complete(&self, buffer: &'static mut [u8], status: Result<(), i2c::Error>) {
+        let nmea_buf = self.sentence_buffer.take().unwrap();
+        let i2c_buf_len = buffer.len();
+
+        if let Err(e) = status {
+            self.i2c_buffer.replace(buffer);
+
+            self.client.map(|call| {
+                call.callback(nmea_buf, 0, Err(e.into()));
+            });
+
+            return;
+        }
+
+        let string = match str::from_utf8(buffer) {
+            Ok(utf8) => utf8,
+            Err(_e) => {
+                self.i2c_buffer.replace(buffer);
+
+                self.client.map(|call| {
+                    call.callback(nmea_buf, 0, Err(ErrorCode::NOSUPPORT));
+                });
+                return;
+            }
+        };
+        let mut nmea_offset = self.nmea_offset.get();
+
+        if let Some(location) = string.find('$') {
+            // This is the start of a new sentence
+
+            if (nmea_offset + location) >= nmea_buf.len() {
+                // We will overflow our buffer, just drop the packet and try again
+                nmea_offset = 0;
+            }
+
+            // First copy the remainder of the sentence to the buffer
+            nmea_buf[nmea_offset..(nmea_offset + location)]
+                .copy_from_slice(string[0..location].as_bytes());
+            nmea_offset += location;
+
+            // Now parse the sentence
+            let sentence = match str::from_utf8(&nmea_buf[0..nmea_offset]) {
+                Ok(utf8) => utf8,
+                Err(_e) => {
+                    self.i2c_buffer.replace(buffer);
+
+                    self.client.map(|call| {
+                        call.callback(nmea_buf, 0, Err(ErrorCode::NOSUPPORT));
+                    });
+                    return;
+                }
+            };
+
+            if sentence.starts_with('$') {
+                // At this point we have a sentence with a `$` at the start.
+                // We report it back to the caller.
+                // We loose the rest of the data we just read though
+                self.i2c_buffer.replace(buffer);
+                self.nmea_offset.set(0);
+
+                self.client.map(|call| {
+                    call.callback(nmea_buf, nmea_offset, Ok(()));
+                });
+
+                return;
+            } else {
+                // The sentence didn't start with `$`. This usually occurs
+                // if we start reading mid-sentence. So we just try again and
+                // get the next one.
+                nmea_offset = 0;
+
+                // Copy the rest of the data
+                let size = i2c_buf_len - location;
+                nmea_buf[nmea_offset..(nmea_offset + size)]
+                    .copy_from_slice(string[location..].as_bytes());
+                nmea_offset += size;
```

Comment:
Is `string` guaranteed to be `i2c_buf_len` long? Can this be rewritten so we can guarantee that copy_from_slice won't panic?

In general this code is concerning because it isn't clear when the index is into a utf8 str and when it is into a u8 buffer, and mixing those will work until it doesn't.